### PR TITLE
[distributions] KL-Divergence for Multivariate Normal

### PR DESCRIPTION
**Pull request by vishwakftw**
This PR implements the KL-Divergence for Multivariate Normal and tests for the same.

Math reviewed and approved by @fritzo [here](https://github.com/probtorch/pytorch/pull/144).

### CODE REVIEW

**Comment by fritzo** [comment]
> @vishwakftw  The tests are failing in the CI. However, the tests are not failing locally.

1. Are you sure the tests are running locally?
2. Try merging master into your branch?
3. Try examining the build log and exactly reproducing the environment it builds (maybe there it is using a different lapack or sth).

File: torch/distributions/kl.py [hunk 1]
```diff
@@ -278,6 +307,21 @@ def _kl_laplace_laplace(p, q):
...
+@register_kl(MultivariateNormal, MultivariateNormal)
+def _kl_multivariatenormal_multivariatenormal(p, q):
+    # From https://en.wikipedia.org/wiki/Multivariate_normal_distribution#Kullback%E2%80%93Leibler_divergence
+    if p.event_shape != q.event_shape:
+        raise ValueError("KL-divergence between two Multivariate Normals with\
+                          different event shapes cannot be computed")
+
+    term1 = _batch_diag(q.scale_tril).log().sum(-1) - _batch_diag(p.scale_tril).log().sum(-1)
+
+    term2 = _batch_trace(_batch_mm(_batch_inverse(q.covariance_matrix), p.covariance_matrix))
```
Comment (vishwakftw) [inline torch/distributions/kl.py]: I am using this is instead of `q.precision_matrix`. Something seems to have been broken between current HEAD and https://github.com/pytorch/pytorch/commit/c2703aa1412fa07f2dc909abda635770a5c62792 .

Comment (fritzo) [reply torch/distributions/kl.py]: @fehiepsi do you think your #6128 may be related?

Comment (fehiepsi) [reply torch/distributions/kl.py]: In my opinion, it is better to use `q.scale_tril` to calculate things related to the inverse of a positive definite matrix.

In this case, we can simplify the computation by using properties of trace:
$ tr( inv(LLt) @ SSt) = tr( inv(Lt) inv(L) S St) = tr( inv(L) S St inv(Lt)) $
So this trace term can be calculated from inv(L) S by squaring it and take sum.

File: torch/distributions/kl.py [hunk 2]
```diff
@@ -278,6 +308,21 @@ def _kl_laplace_laplace(p, q):
...
+
+    term3 = _batch_mahalanobis(q.scale_tril, (q.loc - p.loc))
+    return term1 + 0.5 * (term2 + term3 - p.event_shape[0])
```
Comment (fehiepsi) [inline torch/distributions/kl.py]: I guess here we should have `0.5 * term1` ?

**Comment by vishwakftw** [comment]
@fritzo @fehiepsi There seems to be a problem in the `precision_matrix`. Sample code below:
```
>>> import torch as t
>>> a = t.randn(4, 4)
>>> a = t.distributions.transform_to(t.distributions.constraints.lower_cholesky)(a)
>>> m = t.distributions.MultivariateNormal(loc=t.randn(4), scale_tril=a)
>>> m.covariance_matrix

 1.9258  0.3293 -0.0941 -0.6035
 0.3293  0.2726  0.0188  0.2138
-0.0941  0.0188  0.1522  0.0935
-0.6035  0.2138  0.0935  0.6917
[torch.FloatTensor of size (4,4)]

>>> m.precision_matrix

 2.1758 -0.1910 -0.0405 -0.0837
-0.1910  0.6863  0.0517  0.1311
-0.0405  0.0517  0.1432  0.0066
-0.0837  0.1311  0.0066  0.0370
[torch.FloatTensor of size (4,4)]

>>> m.covariance_matrix.mm(m.precision_matrix)

 4.1817 -0.2260 -0.0785 -0.1410
 0.6457  0.1532  0.0049  0.0162
-0.2223  0.0510  0.0272  0.0148
-1.4156  0.3575  0.0535  0.1048
[torch.FloatTensor of size (4,4)]
```

File: torch/distributions/kl.py (hunk 2 above)
Comment (vishwakftw) [reply torch/distributions/kl.py]: I thought the trace would involve fewer computations than the sum. Regarding `0.5 * term1`, @fritzo and I have discussed it in the `probtorch/pytorch` https://github.com/probtorch/pytorch/pull/144.

**Comment by fritzo** [comment]
@fehiepsi, @vishwakftw's implementation worked before he rebased onto #6128, so it's looking like there is a bug in `.precision_matrix` introduced by #6128. Maybe we can fix the bug in this PR, and add @vishwakftw's example above as a unit test. (sorry if I missed a math error when reviewing #6128).

**Comment by vishwakftw** [comment]
I have found the bug in #6128 and fixed it.

**Comment by fritzo** [comment]
@vishwakftw Nice work, I stared at that code for 10 minutes and couldn't see an error :blush: Can you add your example as a unit test?

**Comment by vishwakftw** [comment]
Sure, I will add a test for verifying inverses and Cholesky decompositions.

File: torch/distributions/kl.py (hunk 2 above)
Comment (fehiepsi) [reply torch/distributions/kl.py]: Thank you @vishwakftw ! Got it ^^

**Comment by fehiepsi** [comment]
LGTM, thanks for fixing the bug @vishwakftw ! Just a note that
```
term2 = _batch_mm(_batch_inverse(q.scale_tril), p.scale_tril).pow(2).sum(-1)
```
And I think that you can move `_batch_mm` to the implementation of MultiraviteNormal distribution to simplify precision_matrix computation. But this is not necessary. ^^

**Review by fritzo — APPROVED** [review]
LGTM

**Comment by fritzo** [comment]
@fehiepsi I'm inclined to implement something simple and correct before getting fancy. What you suggest seems incorrect to me:
```py
term2 = _batch_mm(_batch_inverse(q.scale_tril), p.scale_tril).pow(2).sum(-1)  # correct?
```
This is doing an element-wise square, then summing over the last axis. Can you explain why that is equivalent to a matrix-wise `torch.bmm(m, m.transpose(-1, -2)).diag().sum()`?

**Comment by vishwakftw** [comment]
@fritzo I think I have figured a work-around to use only `scale_tril`. I am testing it locally.

**Comment by fehiepsi** [comment]
@fritzo Assume `m = [x1, x2, ..., xn]` where `x1` is a row vector. Then `(m @ m.t()).diag() = [x1.x1t, x2.x2t, ... xn.xnt]`. I guess you concerned about the batch version of this formula? I will test it in that case.

File: torch/distributions/kl.py [hunk 3]
```diff
@@ -128,13 +128,13 @@ def _x_log_x(tensor):
...
-def _batch_trace(bmat):
+def _batch_trace_XXT(bmat):
     """
-    Utility function for calculating the trace of matrices with arbitrary trailing batch dimensions
+    Utility function for calculating the trace of XX^{T} with X having arbitrary trailing batch dimensions
     """
     mat_size = bmat.size(-1)
-    values = torch.stack([torch.trace(M) for M in bmat.reshape((-1, mat_size, mat_size))])
-    return values.view(bmat.shape[:-2])
+    flat_trace = bmat.reshape(-1, mat_size * mat_size).pow(2).sum(-1)
```
Comment (fritzo) [inline torch/distributions/kl.py]: :+1: this looks better, as equivalent to `.pow(2).sum(-1).sum(-1)`

Comment (fritzo) [reply torch/distributions/kl.py]: @fehiepsi thanks for pointing out this simplification. I had to work out the math, and see you were almost right, just needed one more `.sum(-1)` I believe.

File: torch/distributions/multivariate_normal.py [hunk 4]
```diff
@@ -64,6 +64,17 @@ def _batch_inverse(bmat):
     return flat_inv_bmat.view(bmat.shape)
 
 
+def _batch_mm(bmat1, bmat2):
```
Comment (apaszke) [inline torch/distributions/multivariate_normal.py]: Can’t `torch.matmul` handle that?

File: test/test_distributions.py [hunk 5]
```diff
@@ -2623,6 +2631,45 @@ def test_kl_monte_carlo(self):
                 'Actual (analytic): {}'.format(actual),
             ]))
 
+    # Multivariate normal has a separate Monte Carlo based test due to the requirement of random generation of
+    # positive (semi) definite matrices. n is set to 5, but can be increased during testing.
+    def test_kl_multivariate_normal(self):
+        set_rng_seed(0)  # see Note [Randomized statistical tests]
+        n = 5  # Number of tests for multivariate_normal
+        print('Testing KL(MultivariateNormal, MultivariateNormal) using Monte Carlo')
```
Comment (apaszke) [inline test/test_distributions.py]: Please don’t print anything in the tests

File: torch/distributions/multivariate_normal.py (hunk 4 above)
Comment (fehiepsi) [reply torch/distributions/multivariate_normal.py]: `torch.matmul` is slower in my opinion.

Comment (vishwakftw) [reply torch/distributions/multivariate_normal.py]: @apaszke I have fixed this in the latest commit.

**Review by apaszke — APPROVED** [review]
